### DISCUSSION TITLE
Using Full recompute does not decrease vram usage

### DISCUSSION THREAD

**Discussion by JavaZeroo** [NVIDIA/Megatron-LM#1457] [Q&A]
I don't understand why vram hasn't changed anything. But from the speed of training it looks as if full recompute has been enabled

## baseline

```
TP=8
PP=1
DISTRIBUTED_ARGS="
    --nproc_per_node $GPUS_PER_NODE \
    --nnodes $NNODES \
    --node_rank $NODE_RANK \
    --master_addr $MASTER_ADDR \
    --master_port $MASTER_PORT
"
GPT_ARGS="
    --use-mcore-models \
    --tensor-model-parallel-size ${TP} \
    --pipeline-model-parallel-size ${PP} \
    --num-layers 24 \
    --hidden-size 6144 \
    --ffn-hidden-size 16384 \
    --num-attention-heads 48 \
    --group-query-attention \
    --num-query-groups 8 \
    --seq-length 128 \
    --max-position-embeddings 128 \
    --micro-batch-size 1 \
    --global-batch-size 1 \
    --make-vocab-size-divisible-by 1 \
    --lr 1e-6 \
    --lr-decay-style constant \
    --rotary-base 1000000 \
    --train-iters 101 \
    --untie-embeddings-and-output-weights \
    --disable-bias-linear \
    --init-method-std 0.01 \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    --clip-grad 1.0 \
    --adam-beta1 0.9 \
    --adam-beta2 0.95 \
    --initial-loss-scale 1 \
    --no-gradient-accumulation-fusion \
    --use-flash-attn \
    --use-rotary-position-embeddings \
    --position-embedding-type rope \
    --normalization RMSNorm \
    --norm-epsilon 1e-5 \
    --swiglu \
    --no-masked-softmax-fusion \
    --attention-softmax-in-fp32 \
    --no-load-optim \
    --no-load-rng \
    --seed 42 \
    --bf16 \
    --dist-ckpt-strictness log_unexpected \
    --no-masked-softmax-fusion \
    --no-bias-swiglu-fusion \
    --no-bias-dropout-fusion \
    --no-rope-fusion \
    --attention-softmax-in-fp32 \
    --accumulate-allreduce-grads-in-fp32 \
"
```

### vram usage
```
    PID USER DEV     TYPE  GPU        GPU MEM    CPU  HOST MEM Command                                                                                                                           
 121397 root   0  Compute N/A   25932MiB  63%   102%   2036MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121398 root   1  Compute N/A   25932MiB  63%   102%   2055MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121399 root   2  Compute N/A   25932MiB  63%   100%   2036MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121400 root   3  Compute N/A   25932MiB  63%   100%   2036MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121401 root   4  Compute N/A   25932MiB  63%   102%   2037MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121402 root   5  Compute N/A   25932MiB  63%   102%   2036MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121403 root   6  Compute N/A   25932MiB  63%   102%   2037MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 121404 root   7  Compute N/A   25932MiB  63%   100%   2039MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
```

### training log

```
2025-03-08 03:39:44] iteration       93/     101 | consumed samples:           93 | elapsed time per iteration (ms): 283.3 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.598389E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:45] iteration       94/     101 | consumed samples:           94 | elapsed time per iteration (ms): 285.1 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.579762E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:45] iteration       95/     101 | consumed samples:           95 | elapsed time per iteration (ms): 287.5 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.523883E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:45] iteration       96/     101 | consumed samples:           96 | elapsed time per iteration (ms): 282.4 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.402811E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:46] iteration       97/     101 | consumed samples:           97 | elapsed time per iteration (ms): 286.0 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.356245E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:46] iteration       98/     101 | consumed samples:           98 | elapsed time per iteration (ms): 287.5 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.263113E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:46] iteration       99/     101 | consumed samples:           99 | elapsed time per iteration (ms): 290.5 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.263113E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:46] iteration      100/     101 | consumed samples:          100 | elapsed time per iteration (ms): 286.2 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.197920E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:39:47] iteration      101/     101 | consumed samples:          101 | elapsed time per iteration (ms): 285.7 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 2.151354E-07 | loss scale: 1.0 | grad norm: 0.000 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

## recompute

after adding:
```shell
    --recompute-granularity full \
    --recompute-method uniform \
    --recompute-num-layers 24 \
```
### training log

```
 [2025-03-08 03:37:35] iteration       93/     101 | consumed samples:           93 | elapsed time per iteration (ms): 372.5 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.931662E+00 | loss scale: 1.0 | grad norm: 106.125 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:36] iteration       94/     101 | consumed samples:           94 | elapsed time per iteration (ms): 378.3 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.905890E+00 | loss scale: 1.0 | grad norm: 105.842 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:36] iteration       95/     101 | consumed samples:           95 | elapsed time per iteration (ms): 373.4 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.883155E+00 | loss scale: 1.0 | grad norm: 105.635 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:36] iteration       96/     101 | consumed samples:           96 | elapsed time per iteration (ms): 379.1 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.860850E+00 | loss scale: 1.0 | grad norm: 105.362 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:37] iteration       97/     101 | consumed samples:           97 | elapsed time per iteration (ms): 383.2 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.842552E+00 | loss scale: 1.0 | grad norm: 105.141 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:37] iteration       98/     101 | consumed samples:           98 | elapsed time per iteration (ms): 376.8 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.819790E+00 | loss scale: 1.0 | grad norm: 104.850 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:37] iteration       99/     101 | consumed samples:           99 | elapsed time per iteration (ms): 377.7 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.797119E+00 | loss scale: 1.0 | grad norm: 104.588 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:38] iteration      100/     101 | consumed samples:          100 | elapsed time per iteration (ms): 375.4 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.776598E+00 | loss scale: 1.0 | grad norm: 104.269 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
 [2025-03-08 03:37:38] iteration      101/     101 | consumed samples:          101 | elapsed time per iteration (ms): 367.8 | learning rate: 1.000000E-06 | global batch size:     1 | lm loss: 1.756780E+00 | loss scale: 1.0 | grad norm: 104.023 | num zeros: 0 | number of skipped iterations:   0 | number of nan iterations:   0 |
```

### vram usage

```
   PID USER DEV     TYPE  GPU        GPU MEM    CPU  HOST MEM Command                                                                                                                           
 116035 root   7  Compute N/A   24438MiB  60%    32%   3248MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116028 root   0  Compute N/A   24722MiB  60%    45%   3249MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116032 root   4  Compute N/A   24726MiB  60%    26%   3251MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116034 root   6  Compute N/A   24726MiB  60%    42%   3248MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116030 root   2  Compute N/A   24918MiB  61%    39%   3251MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116031 root   3  Compute N/A   25206MiB  62%    27%   3272MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116029 root   1  Compute N/A   25210MiB  62%    31%   3270MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
 116033 root   5  Compute N/A   25210MiB  62%    30%   3268MiB /usr/bin/python -u pretrain_gpt.py --use-mcore-models --tensor-model-parallel-size 8 --pipeline-model-parallel-size 1 --num-layers
```

**Comment by Advait251206** [comment]
There are two things happening here.

First, your results actually show that activation recomputation **is** working:

| Configuration | Iteration Time | GPU Memory |
|---------------|---------------:|-----------:|
| Baseline | ~285 ms | 25.9 GB |
| Full recompute | ~375 ms | 24.4–25.2 GB |

The iteration time increases by roughly **30%**, which is exactly what we expect when full activation recomputation is enabled. During the backward pass, activations are recomputed instead of being read from memory, so training becomes slower.

However, the memory reduction is only about **0.7–1.5 GB**, which is much smaller than many users expect.

The reason is that **activation memory is only one part of the total GPU memory footprint**. A more complete breakdown is

\[
M =
M_{\text{parameters}}
+
M_{\text{optimizer}}
+
M_{\text{gradients}}
+
M_{\text{activations}}
+
M_{\text{temporary workspaces}}
+
M_{\text{communication buffers}}
+
M_{\text{CUDA allocator}}
\]

Activation recomputation only reduces the **activation** term. Everything else remains essentially unchanged.

---

## Why is the memory reduction so small?

Looking at your configuration:

```bash
--tensor-model-parallel-size 8
--pipeline-model-parallel-size 1
--micro-batch-size 1
--seq-length 128
```

the activation memory is already quite small.

Activation memory roughly scales with

\[
\text{MicroBatch}
\times
\text{Sequence Length}
\times
\text{Hidden Size}
\times
\text{Number of Layers}
\]

Since you are using

- **micro-batch-size = 1**
- **sequence length = 128**

there simply aren't many activations to save.

Most of your 25 GB VRAM is occupied by

- model parameters,
- optimizer states,
- gradients,
- FlashAttention/cuBLAS workspaces,
- NCCL communication buffers,
- CUDA caching allocator reservations.

Those components are unaffected by activation checkpointing.

---

## `nvidia-smi` can also be misleading

Another important point is that `nvidia-smi` reports **reserved GPU memory**, not only actively used memory.

PyTorch's CUDA caching allocator typically keeps memory reserved instead of immediately returning it to CUDA. As a result,

```python
torch.cuda.memory_allocated()
```

may decrease noticeably,

while

```python
torch.cuda.memory_reserved()
```

(or `nvidia-smi`) changes very little.

This often makes it appear as if recomputation "didn't save memory," even though the allocator is simply holding onto previously allocated blocks.

To verify the actual reduction, compare

```python
torch.cuda.max_memory_allocated()
```

instead of relying solely on `nvidia-smi`.

---

## Why training becomes slower

Your timing clearly shows

- Baseline: **~285 ms/iteration**
- Full recompute: **~375 ms/iteration**

This is exactly the expected trade-off.

Instead of storing intermediate activations during the forward pass, Megatron recomputes them during backward, which reduces activation memory but increases computation.

So the slower runtime is actually a good indication that recomputation is active.

---

## When does recomputation provide significant memory savings?

Activation recomputation becomes much more effective when activation memory dominates the overall footprint, for example:

- Larger sequence lengths (e.g., 4096–32768)
- Larger micro-batch sizes
- Larger hidden dimensions
- Models with many transformer layers

With your current configuration (`seq-length=128`, `micro-batch-size=1`), activation memory is already a relatively small fraction of total VRAM, so even eliminating a large portion of it only reduces total memory by about **1 GB**.

---

## Conclusion

Based on the logs you've shared, I don't think recomputation is failing. In fact, the increased iteration time strongly suggests it is working as intended.

The reason you don't observe a dramatic VRAM reduction is that your workload has **very small activation memory** (`micro-batch-size=1`, `seq-length=128`), so the majority of GPU memory is consumed by parameters, optimizer states, gradients, communication buffers, temporary workspaces, and the CUDA caching allocator. Since activation recomputation only reduces the activation component, a memory reduction of roughly **1 GB** is entirely reasonable for this configuration.

---

If this answer helped or pointed you in the right direction, I'd appreciate it if you could mark it as the accepted answer so it's easier for others with the same issue to find.

Also, if you found my contribution useful, I'd appreciate it if you could check out my GitHub profile, follow me, and star any repositories you find interesting.

GitHub: https://github.com/Advait251206
